### PR TITLE
Store player storages and learned spells in binary

### PR DESCRIPTION
**Pull request by danyelvarejao1**
Storages and learned spells are now blobbed in the database, improving performance.
Correction of the isPersistent condition function

### CODE REVIEW

**Comment by EPuncker** [comment]
how do I query a storage value after this?
for example, right now I can do this:

`local resultId = db.storeQuery("SELECT `player_id` FROM `player_storage` WHERE `key` = " .. storage .. " AND `value` >= 3;")`

how would I do after your change? just a question 😛

**Comment by Lyuzera** [comment]
> how do I query a storage value after this?
> for example, right now I can do this:
> 
> `local resultId = db.storeQuery("SELECT `player_idplayer_storagekeyvalue`FROM``WHERE``= " .. storage .. " AND`` >= 3;")`
> 
> how would I do after your change? just a question 😛

inside the tfs or website? at tfs you can get normally with `player:getStorageValue`. I prefer binary because it is much faster than the generic estructure, but to show storages on website is not cool.

**Comment by EPuncker** [comment]
> > how do I query a storage value after this?
> > for example, right now I can do this:
> > `local resultId = db.storeQuery("SELECT `player_idplayer_storagekeyvalue`` FROM``WHERE``= " .. storage .. " AND`` >= 3;") ``
> > how would I do after your change? just a question 😛
> 
> inside the tfs or website? at tfs you can get normally with `player:getStorageValue`. I prefer binary because it is much faster than the generic estructure, but to show storages on website is not cool.

not website, but getting storage of an offline player

**Comment by Lyuzera** [comment]
> not website, but getting storage of an offline player

good question, we lost that possibility, because unserialize is bad for get a simple information. But for me, that possibility can be deprecated.. so why?

```lua
-- we can store anything by player guid and get it from all datapack interfaces (thx lua state)
--  simpleCache[player:getGuid()] = {}

-- if you created a big system, you can use cache lua + database
-- storage is very well to quests, but we no need to get infomations when player is offline.
```

File: data/migrations/25.lua [hunk 1]
```diff
@@ -1,3 +1,8 @@
 function onUpdateDatabase()
-	return false
-end
+	print("> Updating database to version 26 (store player storages and learned spells in binary)")
+	db.query("DROP TABLE `player_storage`;")
+	db.query("DROP TABLE `player_spells`;")
+	db.query("ALTER TABLE `players` ADD `spells` blob DEFAULT NULL AFTER `conditions`;")
+	db.query("ALTER TABLE `players` ADD `storages` mediumblob DEFAULT NULL AFTER `spells`;")
```
Comment (DSpeichert) [inline data/migrations/25.lua:6]: So if someone upgrades to this version all their spells/storage is... gone.

**Comment by EPuncker** [comment]
> > not website, but getting storage of an offline player
> 
> good question, we lost that possibility, because unserialize is bad for get a simple information. But for me, that possibility can be deprecated.. so why?
> 
> ```lua
> -- we can store anything by player guid and get it from all datapack interfaces (thx lua state)
> --  simpleCache[player:getGuid()] = {}
> 
> -- if you created a big system, you can use cache lua + database
> -- storage is very well to quests, but we no need to get infomations when player is offline.
> ```

okay, I was only thinking about small things, like achievements displays, hunted lists, custom points rankings (anything that deals with values of offline players) will need to have a sql table I think... 😢  sacrificing a functionality for performance 🤔 i hope devs find a way

**Comment by nekiro** [comment]
I dont think we need that type of optimization on storages, they are fine and does not need to be optimized. The real optimization is needed in items storing.

**Comment by CristoferMartins** [comment]
Storages never been a performance problem even in servers which uses a lot of them. Items storage(specially inbox) is the problem. And btw the ideal fix is to load/save players in another thread in an async fashion for the dispatcher.

**Comment by danyelvarejao1** [comment]
@gesior optimized the saving of items/depot items/inbox items in binary and it is perfect, it has been 100% tested on production servers and the PR of it has never been approved.

**Comment by Lyuzera** [comment]
The @gesior's binary PR was really 100% tested and approved. I've never seen anything as fast as binary items, 150k items and no lag during the login.. and it is faster than cache in RAM.

**Comment by DSpeichert** [comment]
This PR causes data loss and as such has no change of being merged. Let me know if you intend to work on that anymore or if you'd rather close it.

**Comment by danyelvarejao1** [comment]
I see no reason for so much "data loss", in most cases who is using the repository is who is creating/developing your server. Whoever has an online server and with many players does not care for the current repository, you sometimes do not give a damn to those who are updating/optimizing codes that are horrible in TFS just because there is a "data loss", I am referring to @gesior code.
You can actually close the pull request, I also agree that storages should not be changed to binary, but think about it, value people who want to contribute to the repository and improve various things.

**Pull request closed without merging**